src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -843,137 +999,205 @@ void Replayer<I>::create_mirror_snapshot(
-        }
-        // Makesure the image snapshot is COMPLETE
-        if (mirror_ns->group_snap_id == group_snap_id && mirror_ns->complete) {
-          image_snap_complete = true;
-          cls::rbd::ImageSnapshotSpec snap_spec;
-          snap_spec.pool = image.spec.pool_id;
-          snap_spec.image_id = image.spec.image_id;
-          snap_spec.snap_id = snap_info.id;
-          auto it = std::find_if(
-            itl->snaps.begin(), itl->snaps.end(),
-            [&snap_spec](const cls::rbd::ImageSnapshotSpec &s) {
-              return snap_spec.pool == s.pool && snap_spec.image_id == s.image_id;
-            });
-          if (it == itl->snaps.end()) {
-            local_image_snap_specs.push_back(snap_spec);
-          }
-          continue;
-        } else {
-          dout(10) << "remote group snap id: " << group_snap_id
-                   << ", local reflected in the image snap: "
-                   << mirror_ns->group_snap_id << dendl;
-        }
+      auto mirror_ns = std::get_if<cls::rbd::MirrorSnapshotNamespace>(
+          &snap_info.snapshot_namespace);
+      if (!mirror_ns) {
+        continue;
       }
 
-      // In case if the image replayer is not yet updated with the limits, do it here
-      if (!image_snap_complete) {
-        set_image_replayer_limits(image.spec.image_id, &(*itr), locker);
+      // make sure the image snapshot is COMPLETE
+      if (mirror_ns->group_snap_id == group_snap_id && mirror_ns->complete) {
+        image_snap_complete = true;
+        cls::rbd::ImageSnapshotSpec snap_spec;
+        snap_spec.pool = image.spec.pool_id;
+        snap_spec.image_id = image.spec.image_id;
+        snap_spec.snap_id = snap_info.id;
+
+        // check if this spec already exists in local snaps
+        auto it = std::find_if(local_snap.snaps.begin(), local_snap.snaps.end(),
+          [&snap_spec](const cls::rbd::ImageSnapshotSpec &s) {
+            return snap_spec.pool == s.pool &&
+                   snap_spec.image_id == s.image_id;
+          });
+        if (it == local_snap.snaps.end()) {
+          local_image_snap_specs.push_back(snap_spec);
+        }
+        continue;
+      } else {
+        dout(10) << "remote group snap id: " << group_snap_id
+                 << ", local reflected in the image snap: "
+                 << mirror_ns->group_snap_id << dendl;
       }
-      image_snap_complete = false; // reset
     }
-  } else {
-    derr << "remote group snapshot doesnt exist: " << group_snap_id << dendl;
-    on_finish->complete(-ENOENT);
-    return;
+
+    if (!image_snap_complete) {
+      set_image_replayer_limits(image.spec.image_id, &remote_snap, &locker);
+    }
   }
 
-  if (itr->snaps.size() == local_image_snap_specs.size()) {
-    itl->snaps = local_image_snap_specs;
-    itl->state = cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE;
+  cls::rbd::GroupSnapshot local_snap_copy;
+  if (remote_snap.snaps.size() == local_image_snap_specs.size()) {
+    local_snap_copy = local_snap;
```

Comment:
A default-initialized `local_snap_copy` ends up being used by `dout(10)` below:
```suggestion
  auto local_snap_copy = local_snap;
  if (remote_snap.snaps.size() == local_image_snap_specs.size()) {
```

Reply:
There is a return in the else block, so this won't happen and hence it should be fine as it is @idryomov.

Reply:
Ah, I missed that.  IMO it should be made more obvious by moving the dout into the if branch and make `local_snap_copy` variable local to that branch.